### PR TITLE
CI: standardize job names

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-20480

#### What does this implement/fix?
~~While working on gh-20480, I would have found it helpful if the slow test durations were added for all CI jobs, since the lists can be quite different on different platforms.  For example, on `Linux Meson tests / Python-debug & Atlas`, the slowest test by far (22.61s) is `interpolate/tests/test_rgi.py::TestRegularGridInterpolator::test_descending_points_nd[5-<lambda>-slinear]`, but this doesn't even show up in the top ten lists of other jobs. This PR adds slow test duration output to relevant CI jobs.~~

#### Additional information
~~Instead of just listing the top 10, shall we add the durations of all tests above a certain threshold with [`--durations-min`](https://docs.pytest.org/en/7.1.x/how-to/usage.html?highlight=durations#profiling-test-execution-duration)? The downside is that this adds lines that not everyone is interested in. If that is a concern, perhaps the preceding 1000 progress lines should be the first to go?~~

Many tests include "meson" in the name; e.g. `macos_meson.yml` defines `macOS tests (meson)` with one job called `Meson build` while the other is `M1 test - openblas`. IIUC, SciPy is always built with Meson now, so is use of Meson a distinguishing characteristic of a job, or is there something else that might be more informative? I ask mostly because I think another important piece of information to include consistently in job names is whether they are for the full test suite or just those not marked `slow`. (The Windows tests are marked `full` and `fast`, for instance.) In the past, there have been several times I've wanted to know whether slow tests are being run in a given job, and having this information in the job names would have helped. Concisely noting the Python and NumPy version (e.g. cp311/np1.X or some other way to denote the latest minor version) would also be valuable to include alongside the other characteristics.

This PR now standardizes job names. See https://github.com/scipy/scipy/pull/20482#pullrequestreview-2006751825.

### CODE REVIEW

**Comment by lucascolley** [comment]
> perhaps the preceding ~1000 progress lines should be the first to go?

Sounds like a good idea

**Comment by rgommers** [comment]
> perhaps the preceding ~1000 progress lines should be the first to go?

Unless things have changed, this may not be possible. The one `.` per test keeps CI from aborting due to "no output received" IIRC.

> so is use of Meson a distinguishing characteristic of a job, or is there something else that might be more informative?

+1 all the "meson" parts of file and job names were useful during the transition phase, but now that `setup.py` is gone there is no need for that anymore. Your suggestions sound good.

**Comment by mdhaber** [comment]
Ok, I'll try to adjust the names.

I assume you want the `--durations` and `--timeouts` taken out of this based on our other conversation, Ralf?

If so, that's fine. In that case, I will stop the jobs from installing pytest-timeout of if it is not being used, and I'll only apply `--durations` to the tests that I end up using for `pytest-fail-slow`.

**Comment by rgommers** [comment]
> I assume you want the `--durations` and `--timeouts` taken out of this based on our other conversation, Ralf?

I'd suggest adding it to a representative subset only. If you really want to go for this, please go ahead - however my preference is to not have it in every test invocation and CI log indeed.

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
     uses: ./.github/workflows/commit_message.yml
 
   test:
-    name: cp312 (meson) fast
+    name: fast, py3.12/npAny, dev.py
```
Comment (mdhaber) [inline .github/workflows/windows.yml:25]: I didn't find anything particularly notable about this job or the next. Maybe I'll use these for timing and `fail_slow`?

**Review by mdhaber — COMMENTED** [review]
OK, I pivoted this PR to just clean up the file and job names. 

I removed "meson" from names, since it is no longer a distinguishing feature. "meson" still appears in some places that are less visible (e.g. `test_meson:`) than the filename and display name.

I tried to standardize the display names in the format:
<particularly interesting features>, <fast or full>, <Python/NumPy versions>, <uses dev.py or pip+pytest>

I used "Min" to mean minimum supported version, "Dev" to mean installing from GH repo, "Pre" to mean nightly wheels, and "Any" to mean whatever pip or environment.yml happens to install.

Please change/correct anything; I am new to these files.

File: .github/workflows/windows.yml (hunk 1 above)
Comment (rgommers) [reply .github/workflows/windows.yml:25]: Sure, one Windows job for that seems fine - and this one is pretty generic.

**Review by rgommers — APPROVED** [review]
Overall this is an improvement, the added info is helpful. I'll resist the temptation to micro-optimize. Planning to merge this soon, unless there are blocking comments.

**Comment by rgommers** [comment]
In it goes - thanks Matt!

**Pull request merged**